### ISSUE TITLE
False Positives of no-unused-vars on function type decorators

### ISSUE THREAD

**Issue by DongyangStephenChen** [typescript-eslint/typescript-eslint#3188]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

I've seen that this was a problem in previous versions and was supposed to be fixed but It's an issue that is not resolved by having things on the latest packages. When decorating a type as in the example below I am getting the false positive of the no-unused-vars rule. 

```export type TKoreLogger = (doLog: boolean, type: 'log' | 'warn' | 'error' | 'group' | 'groupEnd' | 'table' | 'trace' | 'time' | 'timeEnd' | 'clear' | 'dir', log?: unknown) => void;```

```JSON
{
    "env": {
        "es2021": true
    },
    "extends": [
        "airbnb-base",
        "plugin:import/errors",
        "plugin:import/warnings",
        "plugin:import/typescript"
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 12,
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
        "import/extensions": [
           "error",
           "ignorePackages",
           {
             "js": "never",
             "jsx": "never",
             "ts": "never",
             "tsx": "never"
           }
        ]
     }
}

```

**Expected Result**

Expecting the code to pass the test

**Actual Result**
 false positive on no-unused-vars

*console output as below*
error    'doLog' is defined but never used  no-unused-vars
error    'type' is defined but never used   no-unused-vars
error    'log' is defined but never used    no-unused-vars
**Additional Info**

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `^4.18.0` |
| `TypeScript`                | `^4.2.3` |
| `ESLint`                    | `^7.22.0` |
| `node`                      | `15.11.0` |

**Comment by bradzacher** [comment]
> - [x] I have read the FAQ and my problem is not listed.

You ticked this box.. but the first FAQ article has your answer.
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code
